src/liballoc/arc.rs

```diff
@@ -555,8 +555,9 @@ impl<T: ?Sized> Drop for Arc<T> {
         // more than once (but it is guaranteed to be zeroed after the first if
         // it's run more than once)
         let ptr = *self._ptr;
-        // if ptr.is_null() { return }
-        if ptr as *mut u8 as usize == 0 || ptr as *mut u8 as usize == mem::POST_DROP_USIZE {
+        let thin = ptr as *const ();
+
+        if !thin.is_null() && thin as usize != mem::POST_DROP_USIZE {
```

Comment:
Would it make sense to abstract the condition to a function, so that it can be given a name and re-used? The code itself is trivial, but it's still easy to get it wrong. `if already_dropped(ptr) { return; }` (or any other better name) would be easy to read and convey the information about what is going on in both places (currently one of the checks refers to the comment of the other code fragment, which finally explains what is going on).